ext/compressors/lz4/lz4_compress.c

```diff
@@ -121,40 +138,143 @@ lz4_decompress(WT_COMPRESSOR *compressor, WT_SESSION *session,
     size_t *result_lenp)
 {
 	WT_EXTENSION_API *wt_api;
-	char *compressed_data;
+	LZ4_PREFIX prefix;
 	int decoded;
-	size_t src_data_len;
+	uint8_t *dst_tmp;
+
+	(void)src_len;					/* Unused parameters */
 
 	wt_api = ((LZ4_COMPRESSOR *)compressor)->wt_api;
 
-	/* Retrieve compressed length from start of the data buffer. */
-	src_data_len = *(size_t *)src;
-	if (src_data_len + sizeof(size_t) > src_len) {
-		(void)wt_api->err_printf(wt_api,
-		    session,
-		    "lz4_decompress: stored size exceeds buffer size");
-		return (WT_ERROR);
+	/*
+	 * Retrieve the true length of the compressed block and source and the
+	 * decompressed bytes to return from the start of the source buffer.
+	 */
+	memcpy(&prefix, src, sizeof(LZ4_PREFIX));
+
+	/*
+	 * Decompress, starting after the prefix bytes. Use safe decompression:
+	 * we rely on decompression to detect corruption.
+	 *
+	 * Two code paths, one with and one without a bounce buffer. When doing
+	 * raw compression, we compress to a target size irrespective of row
+	 * boundaries, and return to our caller a "useful" compression length
+	 * based on the last complete row that was compressed. Our caller stores
+	 * that length, not the length of bytes actually compressed by LZ4. In
+	 * other words, our caller doesn't know how many bytes will result from
+	 * decompression, likely hasn't provided us a large enough buffer, and
+	 * we have to allocate a scratch buffer.
+	 */
```

Comment:
Thanks for the clarification, @Cyan4973.  Given all of that, I think we should go with what we have for now.

In future, we could consider either your option 1 (`LZ4_decompress_destSize`), or adding a method to WiredTiger's `WT_COMPRESSOR` interface that returns the buffer size required for decompression.  We already have `pre_size` for pre-allocating a buffer for compression, this would be the same for decompression.